### PR TITLE
dgoss: init at 0.3.16

### LINKED ISSUE #121929: Add dgoss
**Project description**
This is a follow-up to issue #121496 goss.

`dgoss` is a bash script wrapper for `goss`. It helps install and run `goss` inside docker containers and together with `goss` it's very handy to run continuous integration tests when building docker images. For example it lets you build a docker image, run a container with it and validate that the image works as expected, that files have the right permission / mode, that some port is open, that some command gives the expected result, etc. `dgoss` is maintained and released from the same Github repo as `goss`.

I have tried packaging it but since it's a bash script and it's only my second contribution to nixpkgs, I'm having a hard time. I've tried using `pkgs.resholvePackage` to replace references to commands, such as `cp`, `chmod` and so on. `dgoss` relies on a `GOSS_PATH` environment variable and falls back on `which goss` to locate the `goss` binary. I imagine in nixpkgs we could replace this line and "hardcode" `goss` as a dependency of `dgoss`. However, I was not able to do so. I need your help :sweat_smile: :pray: 

Here is my attempt so far: https://github.com/Hyzual/nixpkgs/blob/5fab7908d23bde034688ef1331ca21acfcb4ee78/pkgs/tools/misc/dgoss/default.nix

**Metadata**

* homepage URL: https://github.com/aelsabbahy/goss/tree/master/extras/dgoss
* source URL: https://github.com/aelsabbahy/goss/blob/master/extras/dgoss/dgoss
* license: Apache License 2.0 (`asl20`)
* platforms: From the README it looks like it's linux and Darwin

### PR DESCRIPTION
**Pull request by Hyzual**
###### Motivation for this change

Closes #121929

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Hyzual** [comment]
Result of `nixpkgs-review pr 124077` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dgoss</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124077` at c53e332e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>dgoss</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/misc/dgoss/default.nix:28:3:
```
   |
28 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/tools/misc/dgoss/default.nix [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  lib,
+  fetchFromGitHub,
+  resholvePackage,
+  substituteAll,
+  # runtime dependencies
+  bash,
+  coreutils,
+  goss,
+  which,
+}:
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/dgoss/default.nix:12]:
```suggestion
{ lib
, fetchFromGitHub
, resholvePackage
, substituteAll
, bash
, coreutils
, goss
, which
}:
```

File: pkgs/tools/misc/dgoss/default.nix [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      rev = "v${version}";
+      sha256 = "1m5w5vwmc9knvaihk61848rlq7qgdyylzpcwi64z84rkw8qdnj6p";
+  };
+
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    mkdir -p "$out"/bin
+    install extras/dgoss/dgoss "$out"/bin/dgoss
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/dgoss/default.nix:30]:
```suggestion
    install -d extras/dgoss/dgoss $out/bin
```

File: pkgs/tools/misc/dgoss/default.nix [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        (substituteAll {
+          src = ./set-goss-path.sh;
+          inherit goss;
+        })
+      ];
+      fake = {
+        external = [ "docker" ];
+      };
+    };
+  };
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/dgoss/default.nix:48]: What is this doing? Why are we not just using echo to write that one file?

Comment (Hyzual) [reply pkgs/tools/misc/dgoss/default.nix:44]: > What is this doing?

I'm not sure I understand the scope of "this" in your question. Do you mean the `set-goss-path.sh` file ? It just sets an environment variable to the path of `goss` binary.

> Why are we not just using echo to write that one file?

I'm not sure I understand your question. Do you mean the `set-goss-path.sh` file ? I had not thought about using echo to write it, I'll try

**Comment by Hyzual** [comment]
Result of `nixpkgs-review pr 124077` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dgoss</li>
  </ul>
</details>

File: pkgs/tools/misc/dgoss/default.nix (hunk 3 above)
Comment (Hyzual) [reply pkgs/tools/misc/dgoss/default.nix]: I replaced the `prologue` attribute with running `sed` command to add the environment variable. Is that better ?
I don't know if it was the right thing to do but I added it in the install phase. Please let me know if that was a mistake.

File: pkgs/tools/misc/dgoss/default.nix (hunk 2 above)
Comment (Hyzual) [reply pkgs/tools/misc/dgoss/default.nix]: That produced an error: "extras/dgoss/dgoss already exists". I tried with `-D` option

**Comment by Hyzual** [comment]
Thanks for the review !

**Pull request merged**